Support for further query, query plan formatting (issue #1624)

kevinrr888 opened the pull request:
Changed JexlFormattedStringBuildingVisitor.java to have the option of including HTML for the various query elements. Updated the screen.css file to have support for the HTML elements added in JexlFormattedStringBuildingVisitor (this screen.css file is used in the webserver's Query Metrics page). Moved the call of formatting the metrics to a more appropriate location.

kevinrr888 committed:
Support for further query, query plan formatting (issue #1624)

kevinrr888 committed:
Removed unnecessary comments

kevinrr888 committed:
Removed unused @param causing build-and-test failure

ivakegg commented:
I was hoping that you would use a decorator pattern instead of having if (includeHtml) scattered throughout the code.
What that means is that you would have a call to decorator.apply(String text, TYPE type) at the end of each one of those visitor functions.
The decorator would implement your own interface that has an apply(String, TYPE) method.
Then all of the html stuff would be implemented within an HtmlDecorator.
And then the next step would be to create a BashDecorator that would also do the coloring stuff but instead for a bash shell.

kevinrr888 committed:
Changed to use decorators. Added EmptyDecorator, BashDecorator, and HtmlDecorator (issue #1624)

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -25,17 +57,31 @@
+    protected static final String DOUBLE_QUOTE = "\"";
+    
+    // allowed methods for composition. Nothing that mutates the collection is allowed, thus we have:
+    private Set<String> allowedMethods = Sets.newHashSet("contains", "retainAll", "containsAll", "isEmpty", "size", "equals", "hashCode", "getValueForGroup",
```
This list of methods seems a little fragile to me.  I wonder if this list should be maintained in the FunctionSet class instead such that if we add methods there we might have a chance at remembering to update this list.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -88,6 +134,9 @@ private static String formatBuiltQuery(String query) {
      * @return
      */
     private static boolean containsOnly(String str, char ch) {
+        str = str.replaceAll("<[^>]*>", ""); // remove all existing html from the string (if present)
```
Can we move the "remove all existing html" and "remove bash color" stuff into the decorator classes?  That way these two lines would reduce to something like decorator.removeDecorations(str).

ivakegg replied:
Alternatively we could build the decorated and undecorated variants at the same time which would avoid this somewhat fragile decoration removal stuff.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -281,14 +388,372 @@ public Object visit(ASTAndNode node, Object data) {
+        
+        node.jjtGetChild(0).jjtAccept(this, sb);
+        
+        decorator.apply(sb, node);
```
beautiful!

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -281,14 +388,372 @@ public Object visit(ASTAndNode node, Object data) {
+            node.jjtGetChild(i).jjtAccept(this, sb);
+            
+            if (i == 0 || i == 1) {
+                if (decorator instanceof HtmlDecorator) {
```
These pieces of code should be encoded as a separate decorator function.  There should be nothing in this class that does something different based on which decorator is being used.

kevinrr888 committed:
Moved allowedMethods to FunctionalSet, created new decorator function, no longer remove html/bash coloring

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -367,9 +926,18 @@ public static void main(String args[]) {
         }
         try {
-            System.out.println(JexlFormattedStringBuildingVisitor.buildQuery(JexlASTHelper.parseJexlQuery(query)));
+            String echoCommand = "echo " + DOUBLE_QUOTE + buildQueryWithBashFormatting(JexlASTHelper.parseJexlQuery(query)) + DOUBLE_QUOTE;
```
OK, it would be nice if the user could supply a parameter to choose html formatting or plain text formatting.  Additionally perhaps the user could supply their own class to be used for decorating.

ivakegg reviewed (changes requested):
What we also need are unit test cases for all of this great work please.

kevinrr888 committed:
Test cases for HTML and Bash decoration, updated main in JexlFormattedStringBuildingVisitor, alphabetically sorted decorator classes, fixed build-and-test failure (?)

kevinrr888 committed:
Fixed build-and-test failure

kevinrr888 committed:
Added script to run JexlFormattedStringBuildingVisitor from cmd line

kevinrr888 commented:
Added the script to run JexlFormattedStringBuildingVisitor from the command line. 

Script removes line 4 of ingest-libs.sh ("`. ../ingest/ingest-env.sh`") before running it and then adds it back after to avoid error of not finding ingest-passwd.sh. This is probably not the best way to go about this, any ideas of a better way would be great. 

Also, the cd command could probably be changed. How can I get the local datawave directory of the user to use that instead?

ivakegg commented on warehouse/ingest-scripts/src/main/resources/bin/util/prettyPrintQuery.sh:
```diff
@@ -0,0 +1,21 @@
+}
+
+# Delete line 4 of ./ingest-libs.sh: ". ../ingest/ingest-env.sh"
+sed -i '4d' ./ingest-libs.sh
```
I see you are trying to be fancy here.  However normally this would not be an issue.  Please remove the sed line here and below.

kevinrr888 replied:
Okay, it was an issue for me but if it's only on my end then its no problem. I'll delete those and you can let me know if it poses a problem on your end

kevinrr888 committed:
Removed sed commands from script

kevinrr888 committed:
Updated prettyPrintQuery script to work if executed in a different directory

ivakegg committed:
Updated the formatMetrics call to modify the metrics inline instead of
creating new instances.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/BashDecorator.java:
```diff
@@ -0,0 +1,221 @@
+    
+    @Override
+    public void apply(StringBuilder sb, ASTAndNode node, Collection<String> childStrings, boolean needNewLines) {
+        sb.append(String.join(AND_COLOR + " && " + NC + (needNewLines ? NEWLINE : ""), childStrings));
```
Probably not a big deal, but I would have only colored the '&&', not the spaces.  This doesn't matter when you're changing the font color, but it might look odd if we're coloring the background.

kevinrr888 replied:
Yes, good idea. I'll change that and in other places where the white space is colored

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/HtmlDecorator.java:
```diff
@@ -0,0 +1,197 @@
+    
+    @Override
+    public void apply(StringBuilder sb, ASTAdditiveOperator node) {
+        sb.append(String.format("<span class=\"add-op\">%s</span>", node.image));
```
Probably could have saved yourself a few keystrokes by creating a method to create the CSS tag.

e.g.
```
String createSpan(String cssClass, String content) {
   return String.format("<span class=\"%s\">%s</span>", cssClass, content);
}
```

Then you just call `createSpan(css, content)` everywhere. It also might help if we ever decide to add additional formatting that affects every span.  Adding in one place vs. many.

kevinrr888 replied:
Great idea. I'll fix that

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/HtmlDecorator.java:
```diff
@@ -0,0 +1,197 @@
+    
+    @Override
+    public void apply(StringBuilder sb, ASTGENode node) {
+        sb.append("<span class=\"greater-than-equal-op\"> >= </span>");
```
Same comment as before about applying formatting to whitespace.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -1,79 +1,136 @@
+ * new, indented line. Additionally, there is the option to decorate the query with added color for easier readability in terminal or on web pages. Same
+ * functionality as JexlStringBuildingVisitor.java except the query is built in a formatted manner.
  */
 public class JexlFormattedStringBuildingVisitor extends JexlStringBuildingVisitor {
```
It looks like there's a lot of code in here copied directly from JexlStringBuildingVisitor.  Did you ever look into keeping all of this functionality in the original JexlStringBuildingVisitor, and adding configuration to turn on/off different formatting features?  My preference would be to avoid duplicating code if we can.

In the past, I have had to make changes to JexlStringBuildingVisitor to fix query issues that we run into (see all of the special handling throughout for backslashes in string literals).  If/when those kinds of issues arise in the future, my concern is that we'll make the updated in one class but not the other, and the logic will eventually drift between the two StringBuildingVisitors.

kevinrr888 replied:
Okay, yes I see the concern with this. At first JexlFormattedStringBuildingVisitor didn't have much code duplication but with how much I've had to change about it, its come to be similar to JexlStringBuildingVisitor. Do you think I should just try to move all the functionality into one class (JexlStringBuildingVisitor)?

jwomeara replied:
Barring any obstacles I'm not seeing - yes, that is what I would do.

jwomeara commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlFormattedStringBuildingVisitor.java:
```diff
@@ -310,21 +708,139 @@ public Object visit(ASTReferenceExpression node, Object data) {
+     * 
+     * @param metrics
+     */
+    public static void formatMetrics(List<? extends BaseQueryMetric> metrics) {
```
This seems like a weird home for this method.  Should this not just be a method in the ShardTableQueryMetricHandler?  The StringBuilder shouldn't have to concern itself with query metrics.

kevinrr888 replied:
Yes I agree. It was meant to just be a temporary spot but I never got around to changing it. Thanks for pointing this out. I'll move it to ShardTableQueryMetricHandler.

jwomeara reviewed (changes requested):
Looks pretty good.  Just a few comments.

kevinrr888 committed:
No longer color whitespace, cleaned up HtmlDecorator, moved formatMetrics to ShardTableQueryMetricHandler, deleted JexlFormattedStringBuildingVisitor and JexlFormattedStringBuildingVisitorTest: moved all functionality to JexlStringBuildingVisitor and JexlStringBuildingVisitorTest

kevinrr888 committed:
Fix build-and-test failure (?)

ivakegg commented on warehouse/ingest-scripts/src/main/resources/bin/util/prettyPrintQuery.sh:
```diff
@@ -0,0 +1,19 @@
+
+JACKSON_JARS=$(findJacksonJars)
+
+java -cp $CLASSPATH:$JACKSON_JARS datawave.query.jexl.visitors.JexlStringBuildingVisitor
```
I don't see the arguments being passed to the JexlStringBuildingVisitor.  I think adding $@ at the end of this line would do it.

ivakegg replied:
Note that I believe unless overridden by the user, the default should be to format using the bash decorator.

kevinrr888 replied:
Ah I had it so the user types the arguments. I'll change that and I'll make the default use the bash decorator.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlStringBuildingVisitor.java:
```diff
@@ -637,8 +868,76 @@ public Object visit(ASTAssignment node, Object data) {
+    public static void main(String args[]) {
+        String query = null;
+        JexlQueryDecorator decorator = null;
+        
```
This is looking pretty good.  I ask for a little change in how the parameters can be specified.  I am thinking a little closer to how xmllint may be used:
```
JexlStringBuildingVisitor [ --html --bash --help ] {JEXL-FILE(s)}

The JexlStringBuildingVisitor program parses one or more JEXL files, specified on the command
line as JEXL-FILE (or the standard input if no filenames are provided).  It prints various types of
output, depending upon the options selected.  It is useful for detecting errors in the JEXL query
and providing a formatted output.
```

kevinrr888 replied:
Okay, I will work on changing that.

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/collections/FunctionalSet.java:
```diff
@@ -22,6 +22,10 @@
     private static final long serialVersionUID = 8759102050325550844L;
     protected Set<T> delegate;
+    // allowed methods for composition. Nothing that mutates the collection is allowed, thus we have:
+    public static final Set<String> allowedMethods = Collections.unmodifiableSet(Sets.newHashSet("contains", "retainAll", "containsAll", "isEmpty", "size",
```
Thank you for moving there here.  This makes a lot more sense.

kevinrr888 replied:
Yup!

ivakegg commented on warehouse/query-core/src/main/java/datawave/query/jexl/visitors/JexlStringBuildingVisitor.java:
```diff
@@ -185,24 +383,30 @@ public Object visit(ASTOrNode node, Object data) {
             childSB.setLength(0);
         }
-        sb.append(String.join(" && ", childStrings));
+        // If needNewLines is false, we should remove the new lines added to the child strings
```
I will have to say that this one scares me a little.  Since we have mergde the whole formatting thing into this one class (which is not a bad idea), we have to be very careful that the unformatted query path does not change the query.  If one of the nodes contains a term that actually has a newline in it, then I think that needs to be preserved.  So if buildMultipleLines is false, then we should never be adding any new lines in the first place.

kevinrr888 replied:
Good catch! I overlooked this. I will fix it

kevinrr888 committed:
Added requested changes

- Updated prettyPrintQuery.sh to take args from cmd line
- Cleaned up JexlStringBuildingVisitor's needNewLines() logic and no
  longer remove newlines from child strings
- Changed how JexlStringBuildingVisitor's main method is used (more similar to xmllint)
- Updated JexlStringBuildingVisitorTest/added a couple more tests
- Created query-interactive-parens.js for the webservers Query Metrics
  page

ivakegg committed:
rebase fixes

hlgp commented:
Closing for now.  Reopen when revived and deconflicted.
